internal/util/GitCliUtil.go

```diff
@@ -0,0 +1,105 @@
+package util
+
+import (
+	"fmt"
+	"go.uber.org/zap"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/config"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+type GitCliUtil struct {
+	logger *zap.SugaredLogger
+}
+
+func NewGitCliUtil(logger *zap.SugaredLogger) *GitCliUtil {
+	return &GitCliUtil{
+		logger: logger,
+	}
+}
+
+const GIT_ASK_PASS = "/git-ask-pass.sh"
+
+func (impl *GitCliUtil) Fetch(rootDir string, username string, password string) (response, errMsg string, err error) {
+	impl.logger.Debugw("git fetch ", "location", rootDir)
+	cmd := exec.Command("git", "-C", rootDir, "fetch", "origin", "--tags", "--force")
+	output, errMsg, err := impl.runCommandWithCred(cmd, username, password)
+	impl.logger.Debugw("fetch output", "root", rootDir, "opt", output, "errMsg", errMsg, "error", err)
+	return output, errMsg, nil
+}
+
+func (impl *GitCliUtil) Pull(rootDir string, username string, password string, branch string) (response, errMsg string, err error) {
+	impl.logger.Debugw("git pull ", "location", rootDir)
+	cmd := exec.Command("git", "-C", rootDir, "pull", "origin", branch, "--force")
+	output, errMsg, err := impl.runCommandWithCred(cmd, username, password)
+	impl.logger.Debugw("fetch output", "root", rootDir, "opt", output, "errMsg", errMsg, "error", err)
+	return output, errMsg, nil
+}
+func (impl *GitCliUtil) Checkout(rootDir string, branch string) (response, errMsg string, err error) {
+	impl.logger.Debugw("git checkout ", "location", rootDir)
+	cmd := exec.Command("git", "-C", rootDir, "checkout", branch, "--force")
+	output, errMsg, err := impl.runCommand(cmd)
+	impl.logger.Debugw("checkout output", "root", rootDir, "opt", output, "errMsg", errMsg, "error", err)
+	return output, errMsg, nil
+}
+
+func (impl *GitCliUtil) runCommandWithCred(cmd *exec.Cmd, userName, password string) (response, errMsg string, err error) {
+	cmd.Env = append(os.Environ(),
+		fmt.Sprintf("GIT_ASKPASS=%s", GIT_ASK_PASS),
+		fmt.Sprintf("GIT_USERNAME=%s", userName),
+		fmt.Sprintf("GIT_PASSWORD=%s", password),
+	)
+	return impl.runCommand(cmd)
+}
+
+func (impl *GitCliUtil) runCommand(cmd *exec.Cmd) (response, errMsg string, err error) {
+	cmd.Env = append(cmd.Env, "HOME=/dev/null")
+	outBytes, err := cmd.CombinedOutput()
+	if err != nil {
+		exErr, ok := err.(*exec.ExitError)
+		if !ok {
+			return "", "", err
+		}
+		errOutput := string(exErr.Stderr)
+		return "", errOutput, err
+	}
+	output := string(outBytes)
+	output = strings.TrimSpace(output)
+	return output, "", nil
+}
+
+func (impl *GitCliUtil) Init(rootDir string, remoteUrl string, isBare bool) error {
```

Comment:
if this function is called multiple times, this function may fail

Reply:
directory is being cleaned inside `init` function so it should not be an issue.

Reply:
do you suspect any concurrency issue here.

Reply:
yes, even if its not true concurrent it may cause issues for eg
first run:  Init at t1 ---- clone at t2
second run: Init at t2`
this can result in t2 failure

Reply:
for each thread we clone in new directory(randomly generated)  so that should not be an issue.

Reply:
yes then it should not be an issue. It is inside tmp directory?

Reply:
yes `/tmp/abcd/{randomString}`